doc: Add design doc for thin-snapshots

Yuggupta27 opened the pull request:
Add design document to allow the creation of VolumeSnapshots
for PVCs backed by thinly-provisioned logical volumes.

Based on: https://github.com/topolvm/topolvm/pull/442

Signed-off-by: Yug Gupta <yuggupta27@gmail.com>

llamerada-jp commented:
Thank you for your proposal. We will take a moment for reviewing it.

llamerada-jp commented:
@Yuggupta27 cc @sp98 

Could you please confirm the following issues?

- The `snapshot` and `clone` are different features I think. I want to handle those features in separate proposals if possible. Is the `clone` feature included in the dev freeze mentioned below?
https://topolvm.slack.com/archives/CT0AS6UQ1/p1647251609636729?thread_ts=1646817356.890029&cid=CT0AS6UQ1
- I think the scope of this proposal is only the `local snapshot/clone` feature. If this is the case, could you please write this?
- We can only create and restore `snapshot` and `clone` in the same node of the source volume without the remote data copy. Could you please additionally write that we need to change the scheduler to realize it.
- There is [the proposal of snapshot not using thin volume](https://github.com/topolvm/topolvm/pull/385). And remote copy feature may be considered in the future. So I want to avoid making those impossible in this proposal.

Yuggupta27 commented:
Thanks for the review @llamerada-jp !

> @Yuggupta27 cc @sp98
> 
> Could you please confirm the following issues?
> 
> * The snapshot and clone are different features I think. I want to handle those features in separate proposals if possible. Is the clone feature included in the dev freeze mentioned below?

Yes. The snapshot and clone are separate features. Added a separate design doc for the [PVC-PVC clone feature](https://github.com/topolvm/topolvm/pull/447). 

> Is the clone feature included in the dev freeze mentioned below?
>   https://topolvm.slack.com/archives/CT0AS6UQ1/p1647251609636729?thread_ts=1646817356.890029&cid=CT0AS6UQ1

Yes, both Snapshots and Clones features are also targeted for the same deadlines as thin-lv feature.

> * I think the scope of this proposal is only the local snapshot/clone feature. If this is the case, could you please write this?
> * We can only create and restore snapshot and clone in the same node of the source volume without the remote data copy. Could you please additionally write that we need to change the scheduler to realize it.

The Snapshots, Restored volumes from the snapshots, and Volume Clones(PVC-PVC clones); will all be present in the same node as the parent volume. Will mention the same in the doc as well :+1: 

> * There is [the proposal of snapshot not using thin volume](https://github.com/topolvm/topolvm/pull/385). And remote copy feature may be considered in the future. So I want to avoid making those impossible in this proposal.

Sure! The thin-lv snapshot design was proposed with keeping thick snapshot support also in mind by adding `type` parameter in the LogicalVolume CR for snapshots. Will love to have suggestions on improving the design and making it future-ready!

satoru-takeuchi reviewed (changes requested):
FYI, @llamerada-jp will also comment later.

llamerada-jp reviewed (changes requested):
- Could you please create a new section to write that how PV/PVC using snapshot as a datasource to place at the same node to the original PV/PVC.
- Do you know can we restore the snapshot using different StorageClass of source PV/PVC? I counld't find it in KEP or other documents. If it is allowed, the user may point to a different device-class from source PV/PVC. I believe that this issue has a significant impact on this design.
- Do you want to select the device-class to which the snapshot will be created in this proposal? If so, would you please add an example of `VolumeSnapshotClass` to describe how this should be specified?

Yuggupta27 commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,202 @@
+metadata:
+  name: snapcontent-b083470e-8293-47cc-810d-9561bd1754e6 # restore pv name
+  annotations:
+spec:
```
Nice catch! that was an overlook from my side, updated now :+1:

Yuggupta27 commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,202 @@
+* Thin snapshot volumes are not supported across the nodes in a cluster. The snapshot volume must be exclusively activated on only one cluster node.
+## Open Questions
+
+- Should we only allow thick snapshot creation for thick volumes and thin-snapshots for thinly-provsioned volumes?
+  - If yes, then we don't need to specify `type` parameter for snapshot volumes in the LogicalVolume CR; as we can assume thin-snapshots if the parent is a thinly-provisioned volume.
```
Sure, we can keep the `type` parameter

Yuggupta27 commented:
> * Could you please create a new section to write that how PV/PVC using snapshot as a datasource to place at the same node to the original PV/PVC.

Thanks for the suggestion! Added a section for the flow of operations in case of a Restore Volume operation [here](https://github.com/topolvm/topolvm/pull/446/files#diff-552a2fdc08796a742045a5300895bdad82bd500a67da333c995a528fa193b872R106).

> * Do you know can we restore the snapshot using different StorageClass of source PV/PVC? I counld't find it in KEP or other documents. If it is allowed, the user may point to a different device-class from source PV/PVC. I believe that this issue has a significant impact on this design.

Restoring to different storageclass should not be allowed from our side as we intend to create a copy of the parent pvc, which involves having the same parameters as the parent.
Although it is something we can look into in the future, but for now, we should allow the clone/restored pvc to be of the same storageclass as the parent.

> * Do you want to select the device-class to which the snapshot will be created in this proposal? If so, would you please add an example of VolumeSnapshotClass to describe how this should be specified?

Added examples of both VolumeSnapshotClass and VolumeSnapshot. As suggested by you, we can have a field in VolumeSnapshotClass `"topolvm.cybozu.com/restore-mode": "local"` which specifies the Snapshots here is created on the local node, and it also opens up the possibility of having remote snapshots(in different node) in future!

llamerada-jp commented:
> Restoring to different storageclass should not be allowed from our side as we intend to create a copy of the parent pvc, which involves having the same parameters as the parent.
Although it is something we can look into in the future, but for now, we should allow the clone/restored pvc to be of the same storageclass as the parent.

OK. Would you please write this limitation to this proposal?

> Added examples of both VolumeSnapshotClass and VolumeSnapshot. As suggested by you, we can have a field in VolumeSnapshotClass "topolvm.cybozu.com/restore-mode": "local" which specifies the Snapshots here is created on the local node, and it also opens up the possibility of having remote snapshots(in different node) in future!

Is it correct that in this proposal, snapshots will be created only for "the same device-class" on the same node? If it is, I would like to write that to be clear.

Yuggupta27 commented:
> > Restoring to different storageclass should not be allowed from our side as we intend to create a copy of the parent pvc, which involves having the same parameters as the parent.
> > Although it is something we can look into in the future, but for now, we should allow the clone/restored pvc to be of the same storageclass as the parent.
> 
> OK. Would you please write this limitation to this proposal?
> 
> > Added examples of both VolumeSnapshotClass and VolumeSnapshot. As suggested by you, we can have a field in VolumeSnapshotClass "topolvm.cybozu.com/restore-mode": "local" which specifies the Snapshots here is created on the local node, and it also opens up the possibility of having remote snapshots(in different node) in future!
> 
> Is it correct that in this proposal, snapshots will be created only for "the same device-class" on the same node? If it is, I would like to write that to be clear.

Thanks, will mention that :+1:

Yuggupta27 commented:
@llamerada-jp @satoru-takeuchi @cupnes which of the proposed approaches mentioned in the doc would you guys suggest for the LogicalVolume CRs in case of snapshots?

llamerada-jp commented:
@Yuggupta27 
About scheduler, the following methods were listed internally by us. Would you please consider which options are better in your case? Or could you let me know if you get any other ideas?
1. The users select the same node as the original PV by the node selector. It is easiest to implement but it takes labor for users.
2. Set node selector by mutating webhook. It is useable on public k8s services like EKS/GKE that are not supported scheduler extenders. It may conflict with CD software such as ArgoCD and other operators. So I think requires validation.
3. Update topolvm-scheduler. It may have the highest discuss and  implementation costs. And it is not usable EKS/GKE etc. We want to avoid this option.

Yuggupta27 commented:
> @Yuggupta27 About scheduler, the following methods were listed internally by us. Would you please consider which options are better in your case? Or could you let me know if you get any other ideas?

Hi @llamerada-jp , So setting the Logical Volume CR values to the same as that of parent(NodeName and deviceClass) [here](https://github.com/topolvm/topolvm/pull/446/files#diff-552a2fdc08796a742045a5300895bdad82bd500a67da333c995a528fa193b872R128-R131) will not be enough for asking for lv creation on a particular node(so that it can be on same node as parent)?
For e.g 
```yaml
  },
  Spec: topolvmv1.LogicalVolumeSpec{
            Name:        pvcName,
            # The cloned lv created on the same node as parent.
            NodeName:    sourceVolume.Spec.NodeName,
            # cloned lv created on the same deviceClass as the parent.
            DeviceClass: sourceVolume.Spec.DeviceClass,
            # the size of the restored logical volume is same as the parent.
            Snapshot: {
                type: thin
                dataSource: <string>
                accessType: rw
            }
        },
  ```

Here, `sourceVol` refers to the parent Logical Volume

llamerada-jp commented:
When `CreateVolume` gRPC method of topolvm-controller called, 
`accessibility_requirements` parameter is passed and it contains the selectable nodes topology list. LogicalVolume CR will be created after `CreateVolume` is called. So I guess we may consider scheduling.
https://github.com/topolvm/topolvm/blob/f76b553b2d0bdf2ac71d009a4cd0ec36cd9aea5d/driver/controller.go#L86-L122

I am not confident in this parameter and will survey it a little.

Yuggupta27 commented:
As per my current understanding, when TopoLVM receives a CreateVolume request, we can check if the request has a `Datasource`.
Any CreateVolume request, that is regarding Restoring a snapshot or Creating a clone, has a `dataSource` parameter set in the request; which we can access via:   ` req.VolumeContentSource`. 

In case a volume has a `dataSource`( i.e. `req.VolumeContentSource != nil` ), we can ignore the scheduling intelligence and directly attempt to create the volume on the same node as the parent(as it will fail otherwise).
Also, if the node does not have enough space, we can let the request fail with appropriate error messages.

Apart from Restored or Cloned volumes, scheduling intelligence will still behave as intended for all the other volumes.

Thoughts?

llamerada-jp commented:
The CSI driver's behavior is specified on CSI spec. When `accessible_topology` is specified, TopoLVM must obey this parameter. I think your assumed flow is correct if `accessible_topology` contains the target node or there are no constraints. We should control the `accessible_topology` parameter before the `CreateVolume` method will be called using some scheduler. I am concerned about the mechanisms for realizing it.

Would you please refer to the note of `TopologyRequirement` definitions?
https://github.com/container-storage-interface/spec/blob/b09ca9e903260a0d715c6cb7766417f47b0057c6/spec.md#createvolume

cupnes commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,232 @@
+
+## Metrics and Monitoring
+
+Should be already covered during adding support for `thin` logical volumes. See [thin-provisioning proposal doc](https://github.com/topolvm/topolvm/pull/442).
```
PR #442 has been merged, please change the URL.
```suggestion
Should be already covered during adding support for `thin` logical volumes. See [thin-provisioning proposal doc](docs/proposals/thin-volumes.md).
```

Yuggupta27 replied:
Thanks, updated :+1:

cupnes reviewed (changes requested):
Please correct just a few details. At the moment, I don't think there are any other problems.

Yuggupta27 commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,232 @@
+* An user creates a `VolumeSnapshot` custom resource.
+* The `snapshot-controller` which was watching the `VolumeSnapshot` object, sends a `Create VolumesnapshotContent` req to the api-server.
+* The `external-snapshotter` kubernetes sidecar receives this request, and then sends a `CreateSnapshot` gRPC call to TopoLVM CSI driver.
+* This `CreateSnapshot` request is received by the `topolovm-controller` which then create a `LogicalVolume` CR.
```
Nice catch! Updated now :+1:

Yuggupta27 commented:
> Please correct just a few details. At the moment, I don't think there are any other problems.

Thanks for the suggestions! Updated now :+1:

cupnes commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,231 @@
+# Snapshot & Restore
+
+> Note: Based on thin-provisioning of volumes. Please refer to [thin-provisioning proposal doc](https://github.com/topolvm/topolvm/pull/442).
```
Please correct this as well.
```suggestion
> Note: Based on thin-provisioning of volumes. Please refer to [thin-provisioning proposal doc](docs/proposals/thin-volumes.md).
```

Yuggupta27 replied:
Great catch! Updated :+1:

cupnes reviewed (changes requested):
Thank you. But please make one more correction.

cupnes commented:
@Yuggupta27 Thank you. LGTM. Please confirm @llamerada-jp's comment.

llamerada-jp commented:
@Yuggupta27 
Would you add a limitation or node that the user needs to specify the appropriate node like below when restoring PVC, as we have talked about before?
https://topolvm.slack.com/archives/CT0AS6UQ1/p1648447741324199?thread_ts=1646817356.890029&cid=CT0AS6UQ1

llamerada-jp commented:
@Yuggupta27 
> We can later perform improvements on the same in follow-up PRs if deemed necessary.

I agree with your comment. For this reason, I would like you to update the proposal as follows point.
- The scheduler does not update and the user needs to specify the node when creating the pod currently.
- An example of how to specify the node as you commented.

Yuggupta27 commented:
> * The scheduler does not update and the user needs to specify the node when creating the pod currently.

Thanks for the suggestion! Updated the point and added the example for the same :+1:

Yuggupta27 commented:
> I agree with your comment. For this reason, I would like you to update the proposal as follows point.
> 
> * The scheduler does not update and the user needs to specify the node when creating the pod currently.
> * An example of how to specify the node as you commented.

Thanks for confirming @llamerada-jp ! Updated the PR with the details. 

>@Yuggupta27 Thank you. LGTM. Please confirm @llamerada-jp's comment.

@cupnes  :+1:

llamerada-jp commented:
@Yuggupta27 
LGTM
I'm going to wait to merge this PR until merging the PR for implementation(#463), is that ok?

Yuggupta27 commented:
> @Yuggupta27 LGTM I'm going to wait to merge this PR until merging the PR for implementation(#463), is that ok?

Sure! Thanks!

cupnes commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,291 @@
+`[app.kubernetes.io/name](http://app.kubernetes.io/name): my-pod`.
+
+To schedule the restored pod on the same node,
+the restored pod has the pod affinity set for te source PVC label.
```
typo
```suggestion
the restored pod has the pod affinity set for the source PVC label.
```

Yuggupta27 replied:
Great catch! Updated :+1:

cupnes reviewed (changes requested):
Please make one correction.

nbalacha commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,274 @@
+* **Restoring Snapshots**: Based on a volumesnapshot, a new PVC can be created that contains the contents of the snapshot.
+
+## Introduction
+
```
Please add  a section for Limitations.

Yuggupta27 commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,274 @@
+# Snapshot & Restore
+
+> Note: Based on thin-provisioning of volumes. Please refer to [thin-provisioning proposal doc](thin-volumes.md).
```
Added a warning at the beginning of the doc and a limitation section

Yuggupta27 commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,274 @@
+* The `external-snapshotter` kubernetes sidecar receives this request, and then sends a `CreateSnapshot` gRPC call to TopoLVM CSI driver.
+* This `CreateSnapshot` request is received by the `topolvm-controller` which then create a `LogicalVolume` CR.
+* `topolvm-node` on the target node finds the earlier created `LogicalVolume` CR.
+* `topolvm-node` sends a volume create request to lvmd to create a thin-snapshot from the thin-volume datasource.
```
I believe we can leave this one as is as it already mentions ` from the thin-volume datasource`.

cupnes commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,276 @@
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+    * labelSelector:
```
```suggestion
      - labelSelector:
```

cupnes commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,276 @@
+      requiredDuringSchedulingIgnoredDuringExecution:
+    * labelSelector:
+          matchExpressions:
+        * key:  [app.kubernetes.io/name](http://app.kubernetes.io/name)
```
```suggestion
            - key: app.kubernetes.io/name
```

Yuggupta27 commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,276 @@
+
+Since the a snapshots can only be taken of a BOUND PVC, and volumeBinding is `waitforfirstconsumer`; it is safe to
+assume that the source PVC already has an application POD using it. So while creating a restored PVC from the snapshot of source, we need
+to schedule the pod/app on the same node. To do that, we can make use of pod affinity.
```
Updated :+1:

Yuggupta27 commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,274 @@
+
+Since the a snapshots can only be taken of a BOUND PVC, and volumeBinding is `waitforfirstconsumer`; it is safe to
+assume that the source PVC already has an application POD using it. So while creating a restored PVC from the snapshot of source, we need
+to schedule the pod/app on the same node. To do that, we can make use of pod affinity.
```
Thanks for the suggestion. True, although POD Affinity works is convenient if the source PVC has an application using it, but Node Affinity covers all the use cases.
Updated to use node affinity now

Yuggupta27 commented on docs/proposals/thin-snapshots-restore.md:
```diff
@@ -0,0 +1,274 @@
+on the same node as the source volume.
+
+Since the a snapshots can only be taken of a BOUND PVC, and volumeBinding is `waitforfirstconsumer`; it is safe to
+assume that the source PVC already has an application POD using it. So while creating a restored PVC from the snapshot of source, we need
```
Thanks for the suggestion! Updated to use Node selectors

npu21 commented:
hi @Yuggupta27 ,Is there a release plan for this feature?
